File: java/lib/src/main/java/dev/argon/esexpr/ESExpr.java
package dev.argon.esexpr;


import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import org.eclipse.collections.api.list.primitive.ImmutableByteList;
import org.eclipse.collections.api.list.primitive.ImmutableIntList;
import org.eclipse.collections.api.list.primitive.ImmutableLongList;
import org.eclipse.collections.api.list.primitive.ImmutableShortList;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Represents an ESExpr value.
 */
public sealed interface ESExpr {

	/**
	 * Gets the tag of this expression.
	 * @return The tag.
	 */
	ESExprTag tag();

	/**
	 * A constructor value.
	 * @param constructor The constructor name.
	 * @param args Positional arguments.
	 * @param kwargs Keyword arguments.
	 */
    public static record Constructor(String constructor, ImmutableList<ESExpr> args, ImmutableMap<String, ESExpr> kwargs) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return new ESExprTag.Constructor(constructor);
		}
	}

	/**
	 * A boolean value.
	 * @param b The boolean value.
	 */
    public static record Bool(boolean b) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.BOOL;
		}
	}

	/**
	 * An integer value.
	 * @param n The integer value.
	 */
    public static record Int(BigInteger n) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.INT;
		}
	}

	/**
	 * A string value.
	 * @param s The string value.
	 */
    public static record Str(String s) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.STR;
		}
	}

	/**
	 * A 32-bit floating point value.
	 * @param f The float value.
	 */
	public static record Float16(short f) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.FLOAT16;
		}

		@Override
		public boolean equals(Object obj) {
			return obj instanceof Float16(var other) &&
				f == other;
		}

		@Override
		public int hashCode() {
			return f;
		}

		@Override
		public String toString() {
			return "Float16[f=" + Float.float16ToFloat(f) + "]";
		}
	}

	/**
	 * A 32-bit floating point value.
	 * @param f The float value.
	 */
    public static record Float32(float f) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.FLOAT32;
		}

		@Override
		public boolean equals(Object obj) {
			return obj instanceof Float32(var other) &&
				Float.floatToRawIntBits(f) == Float.floatToRawIntBits(other);
		}

		@Override
		public int hashCode() {
			return Float.floatToRawIntBits(f);
		}
	}

	/**
	 * A 64-bit floating point value.
	 * @param d The double value.
	 */
    public static record Float64(double d) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.FLOAT64;
		}

		@Override
		public boolean equals(Object obj) {
			return obj instanceof Float64(var other) &&
				Double.doubleToRawLongBits(d) == Double.doubleToRawLongBits(other);
		}

		@Override
		public int hashCode() {
			return Long.hashCode(Double.doubleToRawLongBits(d));
		}
	}

	/**
	 * An array of 8-bit values.
	 * @param b The values.
	 */
	public static record Array8(ImmutableByteList b) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.ARRAY8;
		}
	}

	/**
	 * An array of 16-bit values.
	 *
	 * @param b The values.
	 */
	public static record Array16(ImmutableShortList b) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.ARRAY16;
		}
	}

	/**
	 * An array of 32-bit values.
	 *
	 * @param b The values.
	 */
	public static record Array32(ImmutableIntList b) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.ARRAY32;
		}
	}

	/**
	 * An array of 64-bit values.
	 *
	 * @param b The values.
	 */
	public static record Array64(ImmutableLongList b) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.ARRAY64;
		}
	}

	/**
	 * An array of 128-bit values.
	 *
	 * @param b The complex number values as pairs of longs.
	 */
	public static record Array128(ImmutableLongList b) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.ARRAY128;
		}
	}

	/**
	 * A null value.
	 * @param level The level where the null lives. Used to disambiguate nesting of Optional types.
	 */
    public static record Null(BigInteger level) implements ESExpr {
		@Override
		public ESExprTag tag() {
			return ESExprTag.NULL;
		}
	}



	/**
	 * Codec for arbitrary ESExpr values.
	 * @return The codec.
	 */
	public static ESExprCodec<ESExpr> codec() {
		return CODEC;
	}

	/**
	 * Codec for arbitrary ESExpr values.
	 */
	static final ESExprCodec<ESExpr> CODEC = new ESExprCodec<>() {
		@Override
		public ESExprTagSet tags() {
			return new ESExprTagSet.All();
		}

		@Override
		public boolean isEncodedEqual(ESExpr x, ESExpr y) {
			return x.equals(y);
		}

		@Override
		public ESExpr encode(ESExpr value) {
			return value;
		}

		@Override
		public ESExpr decode(ESExpr expr, FailurePath path) throws DecodeException {
			return expr;
		}
	};
}

File: scala/esexpr/shared/src/main/scala/esexpr/ESExpr.scala
package esexpr

import zio.Chunk
import scala.compiletime.asMatchable

enum ESExpr derives CanEqual {
  case Constructor(constructor: String, args: Seq[ESExpr], kwargs: Map[String, ESExpr])
  case Bool(b: Boolean)
  case Int(n: BigInt)
  case Str(s: String)
  case Float16(f: esexpr.Float16)
  case Float16NaN(bits: Short)
  case Float32(f: Float)
  case Float32NaN(bits: scala.Int)
  case Float64(d: Double)
  case Float64NaN(bits: Long)
  case Array8(b: Chunk[Byte])
  case Array16(b: Chunk[Short])
  case Array32(b: Chunk[scala.Int])
  case Array64(b: Chunk[Long])
  case Array128(b: Chunk[Long])
  case Null(level: BigInt)

  def tag: ESExprTag = ESExprTag.fromExpr(this)

  override def equals(obj: Any): Boolean =
    obj.asMatchable match {
      case other: ESExpr => (this, other) match {
        case (Constructor(c1, args1, kwargs1), Constructor(c2, args2, kwargs2)) =>
          c1 == c2 && args1 == args2 && kwargs1 == kwargs2
        case (Bool(b1), Bool(b2)) => b1 == b2
        case (Int(n1), Int(n2)) => n1 == n2
        case (Str(s1), Str(s2)) => s1 == s2
        case (Float16(f1), Float16(f2)) => esexpr.Float16.float16ToRawShortBits(f1) == esexpr.Float16.float16ToRawShortBits(f2)
        case (Float16NaN(bits1), Float16NaN(bits2)) => bits1 == bits2
        case (Float32(f1), Float32(f2)) => ESExpr.compareFloat(f1, f2)
        case (Float32NaN(bits1), Float32NaN(bits2)) => bits1 == bits2
        case (Float64(d1), Float64(d2)) => ESExpr.compareDouble(d1, d2)
        case (Float64NaN(bits1), Float64NaN(bits2)) => bits1 == bits2
        case (Array8(b1), Array8(b2)) => b1 == b2
        case (Array16(b1), Array16(b2)) => b1 == b2
        case (Array32(b1), Array32(b2)) => b1 == b2
        case (Array64(b1), Array64(b2)) => b1 == b2
        case (Array128(b1), Array128(b2)) => b1 == b2
        case (Null(l1), Null(l2)) => l1 == l2
        case _ => false
      }
      case _ => false
    }
}

object ESExpr extends ESExprObjectPlatformSpecific {
  given ESExprCodec[ESExpr]:
    override lazy val tags: ESExprTagSet = ESExprTagSet.All

    override def isEncodedEqual(x: ESExpr, y: ESExpr): Boolean =
      x == y

    override def encode(value: ESExpr): ESExpr = value
    override def decode(expr: ESExpr): Either[ESExprCodec.DecodeError, ESExpr] = Right(expr)
  end given

  inline given ESExprTagSetProvider[ESExpr]:
    override inline def tags: ESExprTagSet = ESExprTagSet.All
  end given
}
